File: dp/c/unit_of_work.c
// Unit of Work Pattern in C

#include <stdio.h>
#include <string.h>

#define MAX_CHANGES 32

typedef struct { int id; char name[32]; int version; } Entity;
typedef enum { CHG_INSERT, CHG_UPDATE, CHG_DELETE } ChangeType;
typedef struct { ChangeType type; Entity entity; } Change;

typedef struct {
    Change changes[MAX_CHANGES];
    int count;
} UnitOfWork;

void uow_init(UnitOfWork* uow) { uow->count = 0; }

void uow_register_new(UnitOfWork* uow, Entity e) {
    printf("[UoW] NEW: id=%d name=%s\n", e.id, e.name);
    uow->changes[uow->count++] = (Change){CHG_INSERT, e};
}

void uow_register_dirty(UnitOfWork* uow, Entity e) {
    printf("[UoW] DIRTY: id=%d name=%s v%d\n", e.id, e.name, e.version);
    uow->changes[uow->count++] = (Change){CHG_UPDATE, e};
}

void uow_register_deleted(UnitOfWork* uow, int id) {
    printf("[UoW] DELETE: id=%d\n", id);
    Entity e = {.id = id};
    uow->changes[uow->count++] = (Change){CHG_DELETE, e};
}

void uow_commit(UnitOfWork* uow) {
    printf("\n[UoW] Committing %d changes...\n", uow->count);
    for (int i = 0; i < uow->count; i++) {
        Change* c = &uow->changes[i];
        switch (c->type) {
            case CHG_INSERT: printf("  INSERT: %s (%d)\n", c->entity.name, c->entity.id); break;
            case CHG_UPDATE: printf("  UPDATE: %s (v%d)\n", c->entity.name, c->entity.version); break;
            case CHG_DELETE: printf("  DELETE: id=%d\n", c->entity.id); break;
        }
    }
    uow->count = 0;
    printf("[UoW] Done.\n");
}

int main(void) {
    UnitOfWork uow;
    uow_init(&uow);

    Entity alice = {1, "Alice", 1};
    Entity bob = {2, "Bob", 1};
    uow_register_new(&uow, alice);
    uow_register_new(&uow, bob);

    Entity alice2 = {1, "Alice Updated", 2};
    uow_register_dirty(&uow, alice2);
    uow_register_deleted(&uow, 2);

    uow_commit(&uow);
    return 0;
}

File: dp/rust/src/unit_of_work.rs
// Unit of Work Pattern in Rust
// Tracks changes to entities and commits them as a single transaction

use std::collections::HashMap;

#[derive(Debug, Clone)]
struct Entity {
    id: u32,
    name: String,
    version: u32,
}

#[derive(Debug)]
enum Change {
    Insert(Entity),
    Update(Entity),
    Delete(u32),
}

struct UnitOfWork {
    changes: Vec<Change>,
    identity_map: HashMap<u32, Entity>,
}

impl UnitOfWork {
    fn new() -> Self {
        UnitOfWork {
            changes: Vec::new(),
            identity_map: HashMap::new(),
        }
    }

    fn register_new(&mut self, entity: Entity) {
        println!("[UoW] Register NEW: {:?}", entity);
        self.identity_map.insert(entity.id, entity.clone());
        self.changes.push(Change::Insert(entity));
    }

    fn register_dirty(&mut self, entity: Entity) {
        println!("[UoW] Register DIRTY: {:?}", entity);
        self.identity_map.insert(entity.id, entity.clone());
        self.changes.push(Change::Update(entity));
    }

    fn register_deleted(&mut self, id: u32) {
        println!("[UoW] Register DELETE: id={}", id);
        self.identity_map.remove(&id);
        self.changes.push(Change::Delete(id));
    }

    fn commit(&mut self) -> Result<(), String> {
        println!("\n[UoW] Committing {} changes...", self.changes.len());
        for change in &self.changes {
            match change {
                Change::Insert(e) => println!("  INSERT: {} ({})", e.name, e.id),
                Change::Update(e) => println!("  UPDATE: {} (v{})", e.name, e.version),
                Change::Delete(id) => println!("  DELETE: id={}", id),
            }
        }
        self.changes.clear();
        println!("[UoW] Committed successfully.");
        Ok(())
    }

    fn rollback(&mut self) {
        println!("[UoW] Rolling back {} changes", self.changes.len());
        self.changes.clear();
    }
}

fn main() {
    let mut uow = UnitOfWork::new();

    uow.register_new(Entity { id: 1, name: "Alice".into(), version: 1 });
    uow.register_new(Entity { id: 2, name: "Bob".into(), version: 1 });
    uow.register_dirty(Entity { id: 1, name: "Alice Updated".into(), version: 2 });
    uow.register_deleted(2);

    uow.commit().unwrap();
}
